### PR TITLE
Add benchmark like grpc-java, #157

### PR DESCRIPTION
**Pull request by patriknw**
Pushed my current state. Doesn't compile yet, but I have converted the server side.
Can be compared with https://github.com/grpc/grpc-java/tree/master/benchmarks

### CODE REVIEW

File: benchmark-java/src/main/java/akka/grpc/benchmarks/qps/AsyncServer.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        });
+  }
+
+  public void shutdown() {
+    if (benchmarkService != null)
+      benchmarkService.shutdown();
+    benchmarkService = null;
+  }
+
+  public static class BenchmarkServiceImpl implements BenchmarkService {
```
Comment (patriknw) [inline benchmark-java/src/main/java/akka/grpc/benchmarks/qps/AsyncServer.java]: It's rather pleasing to compare the implementation of this service with https://github.com/grpc/grpc-java/blob/master/benchmarks/src/main/java/io/grpc/benchmarks/qps/AsyncServer.java#L194

Comment (ktoso) [reply benchmark-java/src/main/java/akka/grpc/benchmarks/qps/AsyncServer.java]: Indeed, much much nicer :-)

**Comment by patriknw** [comment]
I have converted the remaining parts (client and tests). I will run some tests tomorrow and some additional cleanup.

**Comment by patriknw** [comment]
This is ready for review

File: benchmark-java/src/test/java/akka/grpc/benchmarks/driver/LoadWorkerTest.java [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+        .setReqSize(1000)
+        .setRespSize(1000);
+    clientArgsBuilder.getSetupBuilder().getHistogramParamsBuilder()
+        .setResolution(0.01)
+        .setMaxPossible(60000000000.0);
+    assertWorkOccurred(clientArgsBuilder.build());
+  }
+
+  @Test
+  public void runPingPongAsyncClosedLoop() throws Exception {
```
Comment (patriknw) [inline benchmark-java/src/test/java/akka/grpc/benchmarks/driver/LoadWorkerTest.java]: This test sometimes (not always) fails with:
```
[info] - runPingPongAsyncClosedLoop *** FAILED ***
[info]   java.util.concurrent.ExecutionException: io.grpc.StatusRuntimeException: INTERNAL: HTTP/2 error code: PROTOCOL_ERROR
[info] Received Goaway
[info] Received unexpected frame of type RstStreamFrame for stream 3 in state Closed
[info]   at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357)
[info]   at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1915)
[info]   at scala.concurrent.java8.FuturesConvertersImpl$CF.super$get(FutureConvertersImpl.scala:84)
[info]   at scala.concurrent.java8.FuturesConvertersImpl$CF.$anonfun$get$2(FutureConvertersImpl.scala:84)
[info]   at scala.concurrent.BlockContext$DefaultBlockContext$.blockOn(BlockContext.scala:53)
[info]   at scala.concurrent.package$.blocking(package.scala:142)
[info]   at scala.concurrent.java8.FuturesConvertersImpl$CF.get(FutureConvertersImpl.scala:84)
[info]   at akka.grpc.benchmarks.driver.LoadWorkerTest.assertWorkOccurred(LoadWorkerTest.java:221)
[info]   at akka.grpc.benchmarks.driver.LoadWorkerTest.runPingPongAsyncClosedLoop(LoadWorkerTest.java:173)
[info]   at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[info]   ...
[info]   Cause: io.grpc.StatusRuntimeException: INTERNAL: HTTP/2 error code: PROTOCOL_ERROR
```

Could that be a real issue in our implementation?

File: benchmark-java/src/main/java/akka/grpc/benchmarks/driver/LoadClient.java [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.LockSupport;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Implements the client-side contract for the load testing scenarios.
+ */
+class LoadClient {
```
[the remark this answers is not in the document]
Comment (johanandren) [reply benchmark-java/src/main/java/akka/grpc/benchmarks/driver/LoadClient.java]: Not necessarily wrong, but curious, why Java?

File: benchmark-java/src/main/java/akka/grpc/benchmarks/driver/LoadClient.java [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+
+    // Create the histogram recorder
+    recorder = new Recorder((long) config.getHistogramParams().getMaxPossible(), 3);
+  }
+
+  /**
+   * Start the load scenario.
+   */
+  void start() {
+    Runnable r;
```
Comment (johanandren) [inline benchmark-java/src/main/java/akka/grpc/benchmarks/driver/LoadClient.java]: Why is `r` defined outside of the loop? Could be `final Runnable r;` inside of it

File: benchmark-java/src/main/java/akka/grpc/benchmarks/driver/LoadClient.java [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+  private final ExponentialDistribution distribution;
+  private volatile boolean shutdown;
+  private final int threadCount;
+
+  private final ActorSystem system;
+  private final Materializer mat;
+
+  private final BenchmarkServiceClient[] clients;
+  private final Recorder recorder;
+  private ExecutorService fixedThreadPool;
```
Comment (johanandren) [inline benchmark-java/src/main/java/akka/grpc/benchmarks/driver/LoadClient.java]: Could also be final

File: benchmark-java/src/main/java/akka/grpc/benchmarks/qps/AsyncServer.java [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+ */
+public class AsyncServer {
+  private static final Logger log = Logger.getLogger(AsyncServer.class.getName());
+
+  private ActorSystem system;
+  private BenchmarkServiceImpl benchmarkService;
+
+  /**
+   * checkstyle complains if there is no javadoc comment here.
+   */
```
Comment (johanandren) [inline benchmark-java/src/main/java/akka/grpc/benchmarks/qps/AsyncServer.java]: best javadoc ever

File: benchmark-java/README.md [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+More scenarios can be tested with the `akka.grpc.benchmarks.driver.LoadWorker`. See `LoadWorkerTest`.
+
+## Visualizing the Latency Distribution
+
+The QPS client comes with the option `--save_histogram=FILE`, if set it serializes the histogram to `FILE` which
+can then be used with a plotter to visualize the latency distribution. The histogram is stored in the file format
+of [HdrHistogram](http://hdrhistogram.org/). That way it can be plotted very easily using a browser based tool
+like http://hdrhistogram.github.io/HdrHistogram/plotFiles.html. Simply upload the generated file and it will generate
+a beautiful graph for you. It also allows you to plot two or more histograms on the same surface in order two easily
+compare latency distributions.
```
[the remark this answers is not in the document]
Comment (patriknw) [reply benchmark-java/README.md]: yeah, I don't take responsibility for all code here (much copied from grpc-java), but I can fix this

File: benchmark-java/src/main/java/akka/grpc/benchmarks/driver/LoadClient.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (patriknw) [reply benchmark-java/src/main/java/akka/grpc/benchmarks/driver/LoadClient.java]: Benchmarks are taken from grpc-java and I didn't see the point of converting some or all of it to scala. For MVP I think it's enough to have benchmarks in java and later we could consider adding scala implementations of some parts of it, still plugging in to the java benchmark infrastructure.

[the remark this answers is not in the document]
Comment (johanandren) [reply benchmark-java/src/main/java/akka/grpc/benchmarks/driver/LoadClient.java]: Ah, didn't get that, thought it was the parts not from grpc that you had written, but then it makes sense.

[the remark this answers is not in the document]
Comment (patriknw) [reply benchmark-java/src/main/java/akka/grpc/benchmarks/driver/LoadClient.java]: original code is here: https://github.com/grpc/grpc-java/tree/master/benchmarks

File: benchmark-java/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (johanandren) [reply benchmark-java/README.md]: If copied then leave it be!

**Comment by patriknw** [comment]
Refs #157

File: benchmark-java/src/test/java/akka/grpc/benchmarks/driver/LoadWorkerTest.java (hunk 2 above)
Comment (raboof) [reply benchmark-java/src/test/java/akka/grpc/benchmarks/driver/LoadWorkerTest.java]: Yes, seems like a bug: from https://http2.github.io/http2-spec/#StreamStates: "Endpoints MUST ignore WINDOW_UPDATE or RST_STREAM frames received in this state, though endpoints MAY choose to treat frames that arrive a significant time after sending END_STREAM as a connection error". Added https://github.com/akka/akka-http/issues/2087

Comment (patriknw) [reply benchmark-java/src/test/java/akka/grpc/benchmarks/driver/LoadWorkerTest.java]: thanks for digging

**Comment by patriknw** [comment]
I'll cleanup the remaining now

File: benchmark-java/src/test/java/akka/grpc/benchmarks/driver/LoadWorkerTest.java (hunk 2 above)
Comment (patriknw) [reply benchmark-java/src/test/java/akka/grpc/benchmarks/driver/LoadWorkerTest.java]: if it becomes to annoying we have to mark this test with `@Ignore` until fixed

**Comment by patriknw** [comment]
cleanup done, will merge after validation

**Comment by patriknw** [comment]
It's unrelated https://github.com/akka/akka-grpc/issues/219 that is failing

**Pull request merged**